rfcs/0199-basic-config.md

```diff
@@ -0,0 +1,91 @@
+# RFC - Basic Configuration
+
+This RFC proposes a basic configuration system for use in the kernel.
+
+The goal of this system is to allow for a reduction in target-specific code, including potentially removing separate projects for similar targets, such as the lichee-rv, mq-pro, and beepy projects.
+
+This system aims to address current problems, and get us to the next "order of magnitude" of problems (by ignoring many problems we know will be a "later problem").
+
+## In Broad Strokes
+
+### The `config` crate
+
+We would add a new `config` crate that provides helpers for loading configuration at compile time from a `json` or `toml` format.
+
+This crate would also provide helpers for "baking" this data at compile time using the [databake] crate in a `build.rs` file, and compiling into the target as code.
+
+The `config` crate would also provide a `MnemosConfig` type that is generic over two types:
+
+* A `KernelConfig` type, which contains all kernel-defined configuration, provided by the `kernel` crate
+* A `PlatformConfig` type, which contains all target- or family-specific configuration, expected to be provided by any of the defined `platforms` that mnemos supports
+
+This crate uses the generics to avoid circular dependencies between the config crate and kernel/platform crates. It is intended that "user facing" applications of the config crate (primarily in platform crates) is done by providing type aliases to "hide" the generics.
+
+[databake]: https://docs.rs/databake/latest/databake/
+
+### Kernel changes
+
+The kernel will provide a single `KernelConfig` type, which is a struct that contains all service and daemon configuration values. We will add functionality to the kernel to take a provided `KernelConfig` type, and automatically launch the configured services.
+
+This capability will allow most platform targets to avoid repeated boilerplate `kernel.initialize(...)` calls at startup.
+
+### Platform crate changes
+
+Platform crates will be expected to provide a `PlatformConfig` type, containing all drivers or daemons that are specific or applicable to this platform (other than the ones defined in the kernel itself). Similar to the kernel, they should also provide a single call that takes a `KernelConfig` and `PlatformConfig`, and automatically launches all kernel and platform services.
+
+One platform may have many targets: For example, the Allwinner D1 platform has three targets currently: the `lichee-rv` dev board, the `mq-pro` dev board, and the `beepy`.
```

Comment:
it would be nice to see an example of how we plan to abstract over fairly complex stuff like pin configurations in the config system, but that can be added to the RFC in a follow-up.